### ISSUE TITLE
1290 - Pinia

### ISSUE THREAD

**Issue by Ayub-Begimkulov** [type-challenges/type-challenges#27536]
I haven't saw a solution where someone used generic constraints (i.e. `extends ...`). which I think is really important for the real library. 

Here is my solution:

```ts
type AnyObject = Record<string, unknown>
type BaseGetters = Record<string, Function> ;
type BaseActions = Record<string, (...args:any[]) => any>;

type ComputedGetters<G extends BaseGetters> = {
  readonly [K in keyof G]: G[K] extends (...args:any[]) => any ? ReturnType<G[K]> : never;
};

interface StoreOptions<State extends AnyObject, Getters extends BaseGetters, Actions extends BaseActions> {
  id: string;
  state: () => State,
  getters: Getters & ThisType<ComputedGetters<Getters> & Readonly<State>>,
  actions: Actions & ThisType<Actions & State & ComputedGetters<Getters>>
}

type Store<
  State extends AnyObject, 
  Getters extends BaseGetters, 
  Actions extends BaseActions
>= {
  init(): void;
  reset(): true;
} & Actions & State & ComputedGetters<Getters>;

declare function defineStore<
  State extends AnyObject, 
  Getters extends BaseGetters, 
  Actions extends BaseActions
>(store: StoreOptions<State, Getters, Actions>): Store<State, Getters, Actions>;
```

**Comment by watisnu** [comment]
Great work! This solution is almost perfect, however, you're still manually defining `init()` and `reset()` when these should be caller-defined.

Just defining `BaseActions` also as `Record<string, Function>` solves this issue btw.

**Comment by xuejiangping** [comment]
> Great work! This solution is almost perfect, however, you're still manually defining `init()` and `reset()` when these should be caller-defined.
> 
> Just defining `BaseActions` also as `Record<string, Function>` solves this issue btw.

You're quite right, but I want to know why that is? Moreover, when `type BaseGetters = Record<string, (... args:any[]) => any> `, the store will not get the getters attribute,  must let `type BaseGetters = Record<string,Function>`to work properly.

**Comment by xuejiangping** [comment]
> Great work! This solution is almost perfect, however, you're still manually defining `init()` and `reset()` when these should be caller-defined.
> 
> Just defining `BaseActions` also as `Record<string, Function>` solves this issue btw.

@watisnu You're quite right, but I want to know why that is? Moreover, when type `BaseGetters = Record<string, (... args:any[]) => any>` , the store will not get the getters attribute, must let `type BaseGetters = Record<string,Function>`to work properly.

**Comment by namezy** [comment]
type ComputedGetters<TGetter> = {
  [K in keyof TGetter]: TGetter[K] extends () => infer T ? T : never
}
type Options<TState, TGetter, TAction> = {
  id: string
  state: (this: void) => TState
  getters: TGetter &
    ThisType<Readonly<TState> & Readonly<ComputedGetters<TGetter>>>
  actions: TAction &
    ThisType<TState & Readonly<ComputedGetters<TGetter>> & TAction>
}
type Return<TState, TGetter, TAction> = {
  [K in keyof TState]: TState[K]
} & ComputedGetters<TGetter> & {
    [K in keyof TAction]: TAction[K]
  }
declare function defineStore<TState, TGetter, TAction>(
  store: Options<TState, TGetter, TAction>
): Return<TState, TGetter, TAction>

**Comment by namezy** [comment]
```
type ComputedGetters<TGetter> = {
  [K in keyof TGetter]: TGetter[K] extends () => infer T ? T : never
}
type Options<TState, TGetter, TAction> = {
  id: string
  state: (this: void) => TState
  getters: TGetter &
    ThisType<Readonly<TState> & Readonly<ComputedGetters<TGetter>>>
  actions: TAction &
    ThisType<TState & Readonly<ComputedGetters<TGetter>> & TAction>
}
type Return<TState, TGetter, TAction> = {
  [K in keyof TState]: TState[K]
} & ComputedGetters<TGetter> & {
    [K in keyof TAction]: TAction[K]
  }
declare function defineStore<TState, TGetter, TAction>(
  store: Options<TState, TGetter, TAction>
): Return<TState, TGetter, TAction>
